src/core/options-watcher.cpp

```diff
@@ -0,0 +1,134 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2023 Intel Corporation. All Rights Reserved.
+
+
+#include "options-watcher.h"
+#include <proc/synthetic-stream.h>
+
+
+namespace librealsense {
+
+
+options_watcher::options_watcher( std::chrono::milliseconds update_interval )
+    : _update_interval( update_interval )
+    , _updater()
+{
+}
+
+options_watcher::~options_watcher()
+{
+    stop();
+}
+
+void options_watcher::register_option( rs2_option id, std::shared_ptr< option > option, bool update_from_hw )
+{
+    _options[id] = option;
+    _update_from_hw[id] = update_from_hw;
+    _option_values[id] = update_from_hw ? option->query() : option->query(); // TODO - when implemented use option->get_last_known_value();
+
+    if( should_start() )
+        start();
+}
+
+void options_watcher::unregister_option( rs2_option id )
+{
+    _options.erase( id );
+
+    if( should_stop() )
+        stop();
+}
+
+rsutils::subscription
+options_watcher::subscribe( std::function< void( const std::map< rs2_option, std::shared_ptr< option > > & ) > && callback )
+{
+    rsutils::subscription ret = _on_values_changed.subscribe( std::move( callback ) );
+
+    if( should_start() )
+        start();
+
+    return ret;
+}
+
+rsutils::subscription_slot options_watcher::add_callback(
+    std::function< void( const std::map< rs2_option, std::shared_ptr< option > > & ) > && callback )
+{
+    rsutils::subscription_slot ret = _on_values_changed.add( std::move( callback ) );
+
+    if( should_start() )
+        start();
+
+    return ret;
+}
+
+bool options_watcher::remove_callback( rsutils::subscription_slot callback_id )
+{
+    bool ret = _on_values_changed.remove( callback_id );
+
+    if( should_stop() )
+        stop();
+
+    return ret;
+}
+
+bool options_watcher::should_start() const
+{
+    return ! should_stop();
+}
+
+bool options_watcher::should_stop() const
+{
+    return _on_values_changed.size() == 0 || _options.size() == 0;
```

Comment:
I don't like this.
I don't see the need to keep querying if nobody's watching.

Reply:
If no one is watching then `_on_values_changed.size()` should be zero and `should_stop` would return `true`.
Same if no options are registered, there is nothing to notify about...

Reply:
Right, I misread it.

Reply:
I think it would be easier to describe start/stop behavior in the positive instead of the negative... i.e., `should_stop()` should be `! should_start()`?

Reply:
I don't think it matters. At first `should_start` was implemented with it's own logic, but noticing that it was exactly the opposite of `should_stop` I changed.
Actually keeping both logics might be more efficient. What do you prefer?

Reply:
Leave as is.